validator.go

```diff
@@ -19,16 +19,98 @@ package wrp
 
 import (
 	"errors"
+	"fmt"
+	"reflect"
+	"strings"
 
 	"go.uber.org/multierr"
 )
 
 var (
-	ErrInvalidTypeValidator = errors.New("invalid TypeValidator")
-	ErrInvalidValidator     = errors.New("invalid WRP message type validator")
-	ErrInvalidMsgType       = errors.New("invalid WRP message type")
+	ErrorInvalidValidator      = NewValidatorError(errors.New("invalid WRP message type validator"), nil, "")
+	ErrorInvalidMsgType        = NewValidatorError(errors.New("invalid WRP message type"), []string{"Type"}, "")
+	errorInvalidValidatorError = errors.New("invalid validator error")
 )
 
+type ValidatorError struct {
+	// Fields are the relevant fields involved in Err.
+	Fields []reflect.StructField
+
+	// Err is the cause of the error, e.g. invalid message type.
+	// Either Err or Message must be set.
+	Err error
+
+	// Message is a validation message in case the validator wants
+	// to communicate something beyond the Err cause.
+	Message string
+}
+
+// Is reports whether any error in e.err's chain matches target.
+func (ve ValidatorError) Is(target error) bool {
+	return errors.Is(ve.Err, target)
+}
+
+// Unwrap returns the ValidatorError's Error
+func (ve ValidatorError) Unwrap() error {
+	return ve.Err
+}
+
+func (ve ValidatorError) Error() string {
+	var o strings.Builder
+	o.WriteString("Validator error")
+
+	if len(ve.Fields) > 0 {
+		o.WriteString(" [")
+		for i, f := range ve.Fields {
+			if i > 0 {
+				o.WriteRune(',')
+			}
+
+			o.WriteString(f.Name)
+		}
+
+		o.WriteRune(']')
+	}
+
+	if ve.Err != nil {
+		o.WriteString(" err=")
+		o.WriteString(ve.Err.Error())
+	}
+
+	if len(ve.Message) > 0 {
+		o.WriteString(" msg=")
+		o.WriteString(ve.Message)
+	}
+
+	return o.String()
+}
+
+func NewValidatorError(err error, sf []string, m string) ValidatorError {
+	// Either Err or Message must be set.
+	if err == nil && len(m) == 0 {
+		return ValidatorError{Err: errorInvalidValidatorError}
+	}
+
+	var badFields string
+	verr := ValidatorError{Err: err, Fields: make([]reflect.StructField, 0, len(sf)), Message: m}
+	// Fields must exist.
+	for _, f := range sf {
+		ft, ok := messageReflectType.FieldByName(f)
```

Comment:
I realized that my suggestion of using `reflect.StructField` won't work well in the long-term.  That's because we may need to refer to fields in the `payload`, which might be JSON or some other arbitrary format.

We should probably just simplify this and may the fields a `'[]string`.  Sorry I led you astray!